### ISSUE TITLE
Accuracy and Recally yield same results ?

### ISSUE THREAD

**Issue by MagicFeeling** [Lightning-AI/torchmetrics#1717]
## 🐛 Bug

I wonder if the implementation of accuracy and recall being identical is intended to be like this.

### To Reproduce

Steps to reproduce the behavior...

```
import torch
from torchmetrics.classification import MulticlassAccuracy
from torchmetrics.classification import MulticlassRecall
target = torch.tensor([2, 1, 0, 0])
preds = torch.tensor([[0.16, 0.26, 0.58],[0.22, 0.61, 0.17],[0.71, 0.09, 0.20],[0.05, 0.82, 0.13],])

metric1 = MulticlassRecall(num_classes=3,average='none')
metric2 = MulticlassAccuracy(num_classes=3,average='none')
print(metric1(preds, target))
print(metric2(preds, target))
```

Yields the same result for average ('none',None,'micro','macro')

Could it be that the calculation in line 71 and 76 of the file /src/torchmetrics/functional/classification)/accuracy.py
does the recall calculation _safe_divide(tp, tp + fn) instead of accuracy calculation as _safe_divide(tp + tn, tp + tn + fp + fn)

**Comment by SkafteNicki** [comment]
Hi @MagicFeeling, thanks for reporting this issue. I am pretty sure that our implementation is correct as we are testing fairly strictly. Lets look closer:
Your input is 
```python
target = torch.tensor([2, 1, 0, 0])
preds = torch.tensor([2, 1, 0, 1])  # after taking argmax(dim=1)
```
With only 4 samples, accuracy should be therefore easy to calculate by hand:
* average="micro": we are making 1 mistake out of 4 samples = 0.75
* average=None: we are only getting class 0 wrong a single time = [0.5, 1, 1]
* average="macro": take the average over the score for each class = (0.5 + 1 + 1) / 3 = 0.83

For recall, lets instead rely on the implementation from scikit-learn:
![image](https://user-images.githubusercontent.com/24896311/232516618-4b21a047-dc6b-4169-96bd-1f868cb8d42c.png)

Now our implementation
![image](https://user-images.githubusercontent.com/24896311/232517689-c10c7b11-c734-4b11-8d70-f21d79d30354.png)

Everything seems to match. While the internal of the accuracy implementation may look a bit weird, it should be the correct implementation in the multiclass case where TP, FN, TN, FP are defined a bit differently than in the classical binary setting. The fact is that accuracy and recall is very closely related in formulation and will often be the same. For your example, even Precision will have the exact same value for `"micro"` and `"macro"` average in your case:
![image](https://user-images.githubusercontent.com/24896311/232519082-bfa33dcc-6aaf-47d2-9328-10559568bb9c.png)

Closing issue.

**Comment by MagicFeeling** [comment]
@SkafteNicki Thank you a lot for your detailed explanation. I think the issue I have is that I interpreted the multi class classification problem as a binary classification from the point of one hot encoded tensors. As can be seen in the image I calculated the confusion matrix for all classes and the corresponding metric. All those values are calculated by hand. If I compare them to the Torchmetric multi class implementation, just Iou and Recall match. Multi class accuracy is exactly like recall as it calculates the values from the multiclass integer encoding as presented by you. From a multiclass integer perspective is it still possible to take true negative values into account? Also in my training I get the same values for MulticlassRecall and MulticlassAccuracy.

![image](https://user-images.githubusercontent.com/88619492/232618619-094cd0ab-239f-4b27-a2f4-0e6d0e8bee9f.png)

**Comment by SkafteNicki** [comment]
Hi again @MagicFeeling.
I ran your example and can confirm all the partial results with the individual confusion matrices that you get I completely agree with
```python
target = np.array([3,0,0,1,0,2,0,1,1,0,0,0,1,1,0,0,0,2,0,0,0,0,3,3,3])
preds = np.array([0,0,0,0,0,0,0,1,1,1,3,0,0,0,0,2,0,0,1,0,0,0,0,1,3])
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(target, preds)
#array([[10,  2,  1,  1],
#         [ 3,  2,  0,  0],
#         [ 2,  0,  0,  0],
#         [ 2,  1,  0,  1]], dtype=int64)
FP = matrix.sum(axis=0) - np.diag(matrix)
# array([7, 3, 1, 1], dtype=int64)
FN = matrix.sum(axis=1) - np.diag(matrix)
# array([4, 3, 2, 3], dtype=int64)
TP = np.diag(matrix)
# array([10,  2,  0,  1], dtype=int64)
TN = matrix.values.sum() - (FP + FN + TP)
# array([ 4, 17, 22, 20], dtype=int64)
```
But this example shows very well that the classical formula for (TP + TN) / (TP + FN + FP + TN) does not generalize well to the multiclass task. The problem here is the TN value which will always be very high in multiclass because it is very likely that both your predictions and target have a different class when you do the calculation per class. Just look at class 2: you get an accuracy of 0.88 even though you are identifying none of the two instances correctly, that does not sound right.

The core problem is that accuracy is a *global*  metric and there really should only be one way to calculate it as its `"micro"` average way. Just take sklearn as an example: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html their implementation does not have an `average` argument, because there really is only one way to calculate accuracy. That said, in torchmetrics we provide the option of `"macro"` average because being requested to do so and to standardize with other metrics, but that value will be equal.

In general I would recommend that you evaluate `accuracy` using `micro` average and `recall/precision` using `macro` average.

**Comment by MagicFeeling** [comment]
Thank you a lot for your detailed explanation. I was just confused as the results did not match the official description but resulting equally as recall. Maybe there should be a hint on the documentation. Generally I am aware of misleading accuracy results but the calculation should still follow or match the official definition. But thank you again for your time to clarify this.

**Comment by function2-llx** [comment]
@SkafteNicki Thank you for your explanation. I totally agree with you that accuracy should be a "global" metric, and it makes little sense for a specific class under a multi-class classification task since there are too many TNs. 
However, I still encounter situations I want to calculate such class-specific accuracy for multi-class classification problems. E.g., for medical diagnostic problems, due to some medical-specific concerns, users will focus on both sensitivity (i.e., recall or true positive rate = tp / (tp + fn)) and specificity (i.e., true negative rate = tn / (tn + fp)). And the vanilla definition of accuracy ((tp + tn) / (tp + tn + fp + fn)), which is a kind of "weighted average" of sensitivity and specificity, makes more sense in this case.
So, I feel like it is still nice to let users who know what they're doing calculate class-specific accuracy under a multi-class classification task.
